napari/components/_layer_slicer.py

```diff
@@ -0,0 +1,155 @@
+        self._executor: Executor = ThreadPoolExecutor(max_workers=1)
+        self._layers_to_task: Dict[Tuple[Layer], Future] = {}
+        self._futures: Dict[Future, Layer] = {}
+        self.lock = RLock()
+
+    def slice_layers_async(
+        self, layers: Iterable[Layer], dims: Dims
+    ) -> Future[dict]:
+        """This should only be called from the main thread.
+
+        Creates a new task and adds it to the _layers_to_task dict. Cancels
+        all tasks currently pending for that layer.
+
+        Submitting multiple layers at one generates multiple requests (stored in a dict),
+        but only ONE task.
+
+        A slice submitted with multiple layers will only cancel if the same tuple of layers
+        is reubmitted. If a second slice is requested with only one layer from the first
+        slice, it will not be cancelled.
+        """
+        # Cancel any tasks that are slicing a subset of the layers
+        # being sliced now. This allows us to slice arbitrary sets of
+        # layers with some sensible and not too complex cancellation
+        # policy.
+        # existing_task = self._find_existing_task(layers)
+        # # are the tasks unique?
+        # if existing_task:
+        #     LOGGER.debug('Cancelling task for %s', layers)
+        #     existing_task.cancel()
+        # # remove task
+        # del self._layers_to_task[layers]
+
+        # this is terrible:
+        existing_future = []
+        with self.lock:
+            for layer in layers:
+                for future, future_layer in self._futures.items():
+                    if layer == future_layer:
+                        existing_future.append(layer)
+            for future in existing_future:
+                future.cancel()
+
+        # Not all layer types will initially be asynchronously sliceable.
+        # The following logic gives us a way to handle those in the short
+        # term as we develop, and also in the long term if there are cases
+        # when we want to perform sync slicing anyway.
+        requests = {}
+        for layer in layers:
+            # TODO: temporary to prove a point (which currently fails)
+            # import weakref
+            # assert isinstance(layer, weakref.ReferenceType) # this fails
+            if layer._is_async():
+                requests[layer] = layer._make_slice_request(dims)
+            else:
+                layer._slice_dims(dims.point, dims.ndisplay, dims.order)
+        # # create task for slicing of each request/layer
+        # task = self._executor.submit(
+        #     self._slice_layers, requests
+        # )  # TODO: doesn't this submit all the llayer slicing under one thread?
+
+        # new_futures = {future: layer}
+        new_futures = []
+        with self.lock:
+            for layer, request in requests.items():
+                future = self._executor.submit(self._slice_layers, request)
+                future.add_done_callback(self._on_slice_done)
```

Comment:
It's theoretically possible (though unlikely) for the task to finish before we store the future, which with the code right now will cause `_on_slice_done` to return early. I think the simplest fix here is to always store the future before we add the done callback.